src/Cello/view_ViewMap.hpp

```diff
@@ -0,0 +1,395 @@
+  /// CelloView instance stored internally (in other words the pointer to
+  /// data). Such an operation would not work for one of the backends.
+  ///
+  /// @note
+  /// Alternatively, we could return a constant reference, but that's like
+  /// returning a constant reference to a pointer. In reality, that might be
+  /// more advantageous here since a CelloView is larger than a generic ptr.
+  CelloView<T, 3> operator[] (const std::string& key) const noexcept
+  { return at_(key); }
+  CelloView<T, 3> operator[] (std::size_t index) const noexcept
+  { return at_(index); }
+
+  /// Returns a reference to the mapped view associated with the specified
+  /// index/key
+  CelloView<T, 3> at(const std::string& key) const noexcept
+  { return at_(key); }
+
+  /// Checks whether the container holds the specified key
+  bool contains(const std::string& key) const noexcept
+  { return str_index_map_.contains(key); }
+
+  /// Similar to `at`, but a slice of the view ommitting staled values is
+  /// returned by value
+  ///
+  /// @note
+  /// This is deprecated
+  CelloView<T, 3> get(const std::string& key,
+                      int stale_depth = 0) const noexcept
+  { return get_(key, stale_depth); }
+
+  /// Provided to help debug
+  void print_summary() const noexcept;
+
+  std::size_t size() const noexcept { return str_index_map_.size(); }
+
+  /// Return the name of the instance (if it has one)
+  const std::string& name() const noexcept {return name_;}
+
+  /// Returns the length along a given dimension of each contained view
+  ///
+  /// @param dim Indicates the dimension for which we want the shape. This
+  ///    should be 0, 1, or 2
+  ///
+  /// @note
+  /// The program will fail if this method is invoked and the map holds zero
+  /// elements.
+  /// @note
+  /// Consider changing the name to view_shape in the future
+  int array_shape(unsigned int dim) const noexcept
+  { return views_.array_shape(dim); }
+
+  /// Return a new map holding subsections of each view held by the map
+  ///
+  /// @param slc_z, slc_y, slc_x Instance of CSlice that specify that are
+  ///    passed to the `subarray` method of each contained view.
+  ViewMap<T> subarray_map(const CSlice &slc_z,
+                          const CSlice &slc_y,
+                          const CSlice &slc_x,
+                          const std::string& name = "") const;
+
+  /// Utility method offered for debugging purposes to check whether the
+  /// key-order matches expectations.
+  ///
+  /// @param ref The list of reference keys
+  /// @param raise_err When true, this will raise a fatal error if the order
+  ///    doesn't match
+  /// @param allow_smaller_ref When `true`, the key-order is only compared for
+  ///    first `ref.size()`. When `false` (the default), the map must have the
+  ///    same number of keys as `ref`.
+  bool validate_key_order(const std::vector<std::string> &ref,
+			  bool raise_err, bool allow_smaller_ref = false)
```

Comment:
Maybe it makes sense to have a default value for `raise_err` here also?

Reply:
added